dev/com.ibm.ws.classloading/src/com/ibm/ws/classloading/internal/ContainerClassLoader.java

```diff
@@ -1663,28 +1665,22 @@ protected void addLibraryFile(File f) {
         //this area subject to refactor following shared lib rework..
         //ideally the shared lib code will start passing us ArtifactContainers, and it
         //will own the management of the ACF via DS.
-
-        //NASTY.. need to use DS to get the ACF, not OSGi backdoor ;p
-        BundleContext bc = FrameworkUtil.getBundle(ContainerClassLoader.class).getBundleContext();
-        ServiceReference<ArtifactContainerFactory> acfsr = bc.getServiceReference(ArtifactContainerFactory.class);
-        if (acfsr != null) {
-            ArtifactContainerFactory acf = bc.getService(acfsr);
-            if (acf != null) {
-                //NASTY.. using this bundle as the cache dir location for the data file..
-                try {
-                    ArtifactContainer ac = acf.getContainer(bc.getBundle().getDataFile(""), f);
-                    smartClassPath.addArtifactContainer(ac);
-                } catch (NullPointerException e) {
-                    // TODO completed under task 74097
-                    if (tc.isDebugEnabled()) {
-                        Tr.debug(tc, "Exception while adding files to classpath", e);
-                    }
-                    if (tc.isInfoEnabled()) {
+        final BundleContext bc = FrameworkUtil.getBundle(ContainerClassLoader.class).getBundleContext();
+        acf.call(new Consumer<ArtifactContainerFactory>() {
+            @Override
+            public void accept(ArtifactContainerFactory factory) {
+                File dataFile = bc == null ? null : bc.getDataFile("");
```

Comment:
The check for a null a bundle context (`bc`) was added for reliability, yes?  If `bc` were null in the previous solution, it would cause a NPE at line 1671, before the try-catch(NPE).  I was not aware that shutting down the framework would cause the `Bundle.getBundleContext()` to return a null; it's not obvious from the user doc.  If `bc` is null, then method `addLibraryFile()` could immediately emit the info message, or a debug message, rather than invoke `acf.call()`.

Reply:
I updated the PR to (hopefully) make this more clear.